File: src/main/scala/com/abaddon83/cardsGames/burracoGames/commands/PickUpACardFromDeckCmd.scala
package com.abaddon83.cardsGames.burracoGames.commands

import java.util.UUID

import com.abaddon83.cardsGames.burracoGames.ports.GameRepositoryPort
import com.abaddon83.cardsGames.shares.games.GameIdentity
import com.abaddon83.cardsGames.shares.players.PlayerIdentity
import com.abaddon83.libs.cqs.commands.{Command, CommandHandler}

import scala.concurrent.{Await, Future}

case class PickUpACardFromDeckCmd(
                        gameIdentity: GameIdentity,
                        playerIdentity: PlayerIdentity
                        ) extends Command {
  override protected val requestId: UUID = UUID.randomUUID()
}

class PickUpACardFromDeckHandler(
                                 gameRepositoryPort: GameRepositoryPort
                               )
                               (implicit val ec: scala.concurrent.ExecutionContext) extends CommandHandler[PickUpACardFromDeckCmd]{

   override def handleAsync(command: PickUpACardFromDeckCmd): Future[Unit] = {
     for {
       burracoGameStarted <- gameRepositoryPort.findBurracoGameInitialisedTurnStartBy(gameIdentity = command.gameIdentity)
       burracoGameExecution = burracoGameStarted.pickUpACardFromDeck(playerIdentity = command.playerIdentity)
     } yield gameRepositoryPort.save(burracoGameExecution)
  }

  override def handle(command: PickUpACardFromDeckCmd): Unit = {
    import scala.concurrent.duration._
    import scala.language.postfixOps
    val burracoGameStarted =Await.result(gameRepositoryPort.findBurracoGameInitialisedTurnStartBy(gameIdentity = command.gameIdentity), 500 millis)
    val burracoGameExecution = burracoGameStarted.pickUpACardFromDeck(playerIdentity = command.playerIdentity)
    gameRepositoryPort.save(burracoGameExecution)
  }

}

File: src/main/kotlin/com/abaddon83/burracoGame/commandModel/commands/PickUpACardFromDeckCmd.kt
package com.abaddon83.burracoGame.commandModel.commands

import com.abaddon83.burracoGame.commandModel.models.burracoGameExecutions.BurracoGameExecutionTurnBeginning
import com.abaddon83.burracoGame.commandModel.models.burracoGameExecutions.BurracoGameExecutionTurnExecution
import com.abaddon83.burracoGame.readModel.ports.BurracoGameReadModelRepositoryPort
import com.abaddon83.burracoGame.commandModel.models.games.GameIdentity
import com.abaddon83.burracoGame.commandModel.models.players.PlayerIdentity
import com.abaddon83.burracoGame.commandModel.ports.BurracoGameRepositoryPort
import com.abaddon83.utils.cqs.commands.Command
import com.abaddon83.utils.cqs.commands.CommandHandler
import com.abaddon83.utils.logs.WithLog
import kotlinx.coroutines.runBlocking
import org.koin.core.KoinComponent
import org.koin.core.inject
import java.sql.Timestamp
import java.util.*

data class PickUpACardFromDeckCmd(
        val gameIdentity: GameIdentity,
        val playerIdentity: PlayerIdentity,
        override val commandId: UUID,
        override val commandTimeStamp: Timestamp): Command {

    constructor(gameIdentity: GameIdentity,playerIdentity: PlayerIdentity): this(
            gameIdentity = gameIdentity,
            playerIdentity = playerIdentity,
            commandId = UUID.randomUUID(),
            commandTimeStamp = Timestamp(System.currentTimeMillis()))
}

class PickUpACardFromDeckHandler() : CommandHandler<PickUpACardFromDeckCmd>, KoinComponent, WithLog() {

    private val repository: BurracoGameRepositoryPort by inject()
    
    override fun handle(command: PickUpACardFromDeckCmd) {
        executeCmd(command)
    }

    override suspend fun handleAsync(command: PickUpACardFromDeckCmd) {
        executeCmd(command)
    }

    private fun executeCmd(command: PickUpACardFromDeckCmd) {
        when (val game = repository.getById(command.gameIdentity)) {
            is BurracoGameExecutionTurnBeginning -> repository.save(game
                    .pickUpACardFromDeck(playerIdentity = command.playerIdentity))
            else -> warnMsg("GameIdentity ${command.gameIdentity} not found exist")
        }
    }
}